### PR TITLE
General improvements to HTTPS API (Second Edition)

### LINKED ISSUE #1299: Set Min/MAX TLS version and ciphers for requests
#### What would you like to discuss?

How to set Min/MAX TLS version and ciphers for requests?

#### Checklist

- [x] I have read the documentation.

### LINKED ISSUE #1306: Better HTTP(S) Request options documentation
#### What would you like to discuss?

Since we made some modification to the `options`, this statement
```
options
Type: object

Any of the https.request options.
```
from the [documentation](https://github.com/sindresorhus/got#options) is no more valid.

Maybe the individual `https.Request` options should be documented on the got page.

<hr />

Node.JS options (checked ones are those already documented by Got)

From `http.request`
- [x] agent
- [x] auth
- [x] createConnection (not present in README.md)
- [ ] defaultPort
- [x] family
- [x] headers
- [ ] host
- [ ] hostname
- [ ] insecureHTTPParser
- [x] localAddress
- [x] lookup (not present in README.md)
- [ ] maxHeaderSize
- [x] method
- [x] path (might be reviewed)
- [x] port (not present in README.md)
- [x] protocol (not present in README.md)
- [ ] setHost
- [ ] socketPath
- [x] timeout

From `tls.createSecureContext` (documented `https.request`)
- [x] ca
- [x] cert
- [ ] ciphers (WIP #1435)
- [ ] clientCertEngine
- [ ] crl
- [ ] dhparam
- [ ] ecdhCurve
- [ ] honorCipherOrder (WIP #1435)
- [x] key
- [x] passphrase
- [x] pfx (#1364)
- [x] rejectUnauthorized
- [ ] secureOptions
- [ ] secureProtocol
- [ ] servername
- [ ] sessionIdContext
- [ ] highWaterMark
- [x] checkServerIdentity (unducumented by Node.JS)

From `tls.createSecureContext` (but undocumented in `https.request`)
- [ ] minVersion (WIP #1435)
- [ ] maxVersion (WIP #1435)

### PR DESCRIPTION
**Pull request by Giotino**
Following the previous "General improvements to HTTPS API" (#1255) this PR is going to add the remaining HTTPS options described in #1306 (or, at least, most of them).

This PR is still WIP.

Implemented options to date (completed on "code, test and docs"):
- `ciphers`: code, test
- `honorCipherOrder`: code, test
- `minVersion`: code, test
- `maxVersion`: code, test

#### Checklist

- [x] I have read the documentation.
- [ ] I have included a pull request description of my changes.
- [x] I have included some tests.
- [ ] If it's a new feature, I have included documentation updates in both the README and the types.

(Partially) Fixes #1306

Fixes #1299

### CODE REVIEW

File: source/core/index.ts [hunk 1]
```diff
@@ -858,6 +858,20 @@ export interface HTTPSOptions {
 	The passphrase to decrypt the `options.https.key` (if different keys have different passphrases refer to `options.https.key` documentation).
 	*/
 	passphrase?: SecureContextOptions['passphrase'];
+
+	// TODO add comment
+	ciphers?: SecureContextOptions['ciphers'];
```
Comment (sindresorhus) [inline source/core/index.ts:863]: Can we make this support an array of ciphers instead of string with `:`?

File: test/https.ts [hunk 2]
```diff
@@ -5,26 +5,28 @@ import {DetailedPeerCertificate} from 'tls';
...
 import pem = require('pem');
+import tls = require('tls');
 
 const createPrivateKey = pify(pem.createPrivateKey);
 const createCSR = pify(pem.createCSR);
 const createCertificate = pify(pem.createCertificate);
 const createPkcs12 = pify(pem.createPkcs12);
 
-test('https request without ca', withHttpsServer(), async (t, server, got) => {
+const ciphers = tls.getCiphers().map(cipher => cipher.toUpperCase());
```
Comment (sindresorhus) [inline test/https.ts:15]: Would be more readable if you used destructuring here.

**Review by sindresorhus — CHANGES_REQUESTED** [review]

**Comment by sindresorhus** [comment]
Can you fix the merge conflicts?

**Review by szmarczak — CHANGES_REQUESTED** [review]
I just noticed hat I had this review pending for months. OMG.

**Comment by szmarczak** [comment]
@Giotino Can you rebase this PR please? #1667 has been merged, I'm more than happy to answer all your questions if you encounter anything :)

**Comment by Giotino** [comment]
> In order to be safe and not lose everything feel free to create a new PR and close this one. I think it will be easier to start from `master` again.

Yes, I think it's better to close this PR and open a new one, but recently I haven't been able to work on this repo.

**Comment by szmarczak** [comment]
> recently I haven't been able to work on this repo

No worries!

**Pull request closed without merging**